src/adapter/src/coord/timestamp_oracle/postgres_oracle.rs

```diff
@@ -0,0 +1,399 @@
+        let statement = client.prepare_cached(q).await?;
+        let result = client
+            .query_one(&statement, &[&self.timeline, &proposed_next_ts])
+            .await?;
+
+        let write_ts: i64 = result.try_get("write_ts").expect("missing column write_ts");
+        let write_ts = Timestamp::try_from(write_ts).expect("we don't use negative timestamps");
+
+        debug!(write_ts = ?write_ts, "write_ts");
+
+        let advance_to = write_ts.step_forward();
+
+        Ok(WriteTimestamp {
+            timestamp: write_ts,
+            advance_to,
+        })
+    }
+
+    #[tracing::instrument(name = "oracle::peek_write_ts", level = "debug", skip_all)]
+    async fn fallible_peek_write_ts(&self) -> Result<Timestamp, anyhow::Error> {
+        let q = r#"
+            SELECT write_ts FROM timestamp_oracle
+                WHERE timeline = $1;
+        "#;
+        let client = self.get_connection().await?;
+        let statement = client.prepare_cached(q).await?;
+        let result = client.query_one(&statement, &[&self.timeline]).await?;
+
+        let write_ts: i64 = result.try_get("write_ts").expect("missing column write_ts");
+        let write_ts = Timestamp::try_from(write_ts).expect("we don't use negative timestamps");
+
+        debug!(write_ts = ?write_ts, "peek_write_ts");
+
+        Ok(write_ts)
+    }
+
+    #[tracing::instrument(name = "oracle::read_ts", level = "debug", skip_all)]
+    async fn fallible_read_ts(&self) -> Result<Timestamp, anyhow::Error> {
+        let q = r#"
+            SELECT read_ts FROM timestamp_oracle
+                WHERE timeline = $1;
+        "#;
+        let client = self.get_connection().await?;
+        let statement = client.prepare_cached(q).await?;
+        let result = client.query_one(&statement, &[&self.timeline]).await?;
+
+        let read_ts: i64 = result.try_get("read_ts").expect("missing column read_ts");
+        let read_ts = Timestamp::try_from(read_ts).expect("we don't use negative timestamps");
+
+        debug!(read_ts = ?read_ts, "read_ts");
+
+        Ok(read_ts)
+    }
+
+    #[tracing::instrument(name = "oracle::apply_write", level = "debug", skip_all)]
+    async fn fallible_apply_write(&self, write_ts: Timestamp) -> Result<(), anyhow::Error> {
+        let q = r#"
+            UPDATE timestamp_oracle SET write_ts = GREATEST(write_ts, $2), read_ts = GREATEST(read_ts, $2)
+                WHERE timeline = $1;
+        "#;
+        let client = self.get_connection().await?;
+        let statement = client.prepare_cached(q).await?;
+        let write_ts = i64::try_from(write_ts).expect("timestamp does not fit into i64");
+
+        debug!(new_write_ts = ?write_ts, "apply_write");
+
+        let _ = client
+            .execute(&statement, &[&self.timeline, &write_ts])
+            .await?;
+
+        Ok(())
+    }
+}
+
+// A wrapper around the `fallible_` methods that adds operation metrics and
+// (TODO: future work) retries.
```

Comment:
Given that CI is catching panics with this on, I'm inclined to pull the retries into this PR. Should be pretty straightforward?

Reply:
I'll look into that once I'm through with the other things